### PR TITLE
add IN subquery decorrelation support & fix not handle NOT IN bug

### PR DESCRIPTION
**Pull request by 9DemonFox**
this check-in adds IN subquery decorrelation support for #220. And fix not handle NOT IN bug in #224.

### CODE REVIEW

File: qpmodel/ExprSubquery.cs [hunk 1]
```diff
@@ -317,17 +319,19 @@ public override Value Exec(ExecContext context, Row input)
...
             inlist_().ForEach(x => { inlist.Add(x.Exec(context, input)); });
-            return inlist.Exists(v.Equals);
+            var exist_flag = inlist.Exists(v.Equals);
+            return hasNot_ ? !exist_flag : exist_flag;
         }
 
         public override string ToString()
         {
             var inlist = inlist_();
+            string ifnot = hasNot_ ? "not" : "";
```
Comment (zhouqingqing) [inline qpmodel/ExprSubquery.cs:329]: " not"?

File: qpmodel/SQLParser.cs [hunk 2]
```diff
@@ -254,7 +256,16 @@ public override object VisitInSubqueryExpr([NotNull] SQLiteParser.InSubqueryExpr
                     inlist.Add(Visit(v) as Expr);
                 Expr expr = inlist[0];
                 inlist.RemoveAt(0);
-                return new InListExpr(expr, inlist);
+                bool hasNot;
```
Comment (zhouqingqing) [inline qpmodel/SQLParser.cs:259]: bool hasNot = (context.K_NOT() != null)?

Comment (9DemonFox) [reply qpmodel/SQLParser.cs]: fixed

File: qpmodel/ExprSubquery.cs (hunk 1 above)
Comment (9DemonFox) [reply qpmodel/ExprSubquery.cs]: fixed

**Comment by 9DemonFox** [comment]
**fine turning** is for your recomendation.
**Fix NOT IN subquery BUG; Support simple corelated InSubquery** fix part of issue in #224 
**pass the unitTest related to NOT IN in #issue224**  cancel some recomendation and valid UnitTest related to NOT IN

**Comment by 9DemonFox** [comment]
the uncoverd codes are also uncovered in existsToMarkJoin.

File: test/UnitTest.cs [hunk 3]
```diff
@@ -872,6 +873,46 @@ public void TestExistsSubquery()
...
+            QueryOption option = new QueryOption();
+
+            for (int i = 0; i < 2; i++)
+            {
+                option.optimize_.use_memo_ = i == 0;
+                var phyplan = "";
+
+                //many NOT test
+                //sql = "select a1 from a where a2 not not in (1,2)";
+                //NullReferenceException, see test output.it will rightly get syntax error.
```
Comment (zhouqingqing) [inline test/UnitTest.cs:888]: You can capture the error like this:
```c#
 var sql = "select a2,a3 from a union all select b1,b4 from b group by b1;";
            var result = ExecuteSQL(sql); Assert.IsNull(result);
            Assert.IsTrue(TU.error_.Contains("appear"));
```

File: test/UnitTest.cs [hunk 4]
```diff
@@ -872,6 +873,46 @@ public void TestExistsSubquery()
...
+                //NullReferenceException, see test output.it will rightly get syntax error. 
+
+                var sql = "select a1 from a where a2 not in (1,2)";
+                TU.ExecuteSQL(sql, "2", out phyplan, option);
+                Assert.AreEqual(1, TU.CountStr(phyplan, "not in"));
+                sql = "select a1 from a where a2 in (1,2)";
+                TU.ExecuteSQL(sql, "0;1", out phyplan, option);
+                Assert.AreEqual(0, TU.CountStr(phyplan, "not in"));
+
+                // incorelated InSubquery
```
Comment (zhouqingqing) [inline test/UnitTest.cs:897]: non-correlated?

**Comment by zhouqingqing** [comment]
please update the check-in comments to say this check-in adds IN subquery decorrelation support.

File: test/UnitTest.cs (hunk 3 above)
Comment (9DemonFox) [reply test/UnitTest.cs:888]: got it

**Pull request merged**